### PR TITLE
[SPARK-19074][SS][DOCS] Updated Structured Streaming Programming Guide for update mode and source/sink options

### PR DESCRIPTION
**Pull request by tdas**
## What changes were proposed in this pull request?

Updates
- Updated Late Data Handling section by adding a figure for Update Mode. Its more intuitive to explain late data handling with Update Mode, so I added the new figure before the Append Mode figure.
- Updated Output Modes section with Update mode 
- Added options for all the sources and sinks

---------------------------
---------------------------

image

---------------------------
---------------------------
screen shot 2017-01-03 at 6 09 11 pm
screen shot 2017-01-03 at 6 10 00 pm

---------------------------
---------------------------
image
image
image

### CODE REVIEW

File: docs/structured-streaming-programming-guide.md [hunk 1]
```diff
@@ -910,35 +923,37 @@ Here is the compatibility matrix.
   </tr>
   <tr>
     <td colspan="2" valign="middle"><br/>Queries without aggregation</td>
-    <td>Append</td>
+    <td>Append, Update</td>
     <td>
         Complete mode note supported as it is infeasible to keep all data in the Result Table.
```
Comment (zsxwing) [inline docs/structured-streaming-programming-guide.md]: nit: Complete mode **note** supported

**Review by zsxwing — CHANGES_REQUESTED** [review]
For `structured-streaming-watermark-update-mode.png`, in the third table (under 12:20), `12:00 - 12:10 cat 1` should be gray.

~~Do you want to document that without aggregation or window operators, `update` mode is same as `append` mode?~~ Never mind. Just found that `update` mode requires aggregation.

File: docs/structured-streaming-programming-guide.md [hunk 2]
```diff
@@ -910,35 +923,37 @@ Here is the compatibility matrix.
   </tr>
   <tr>
     <td colspan="2" valign="middle"><br/>Queries without aggregation</td>
-    <td>Append</td>
+    <td>Append, Update</td>
```
Comment (zsxwing) [inline docs/structured-streaming-programming-guide.md]: nit: it doesn't support `update`. See https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala#L76

**Comment by tdas** [comment]
Good catch about non-aggregation queries. we should support update mode, which is same as append mode. I will fix that in a follow up PR.

File: docs/structured-streaming-programming-guide.md [hunk 3]
```diff
@@ -433,6 +433,51 @@ In Spark 2.0, there are a few built-in sources.
...
+  <tr>
+    <td><b>File source</b></td>
+    <td>
+        <code>path</code>: path to the input directory, and common to all file formats.
+        <br/><br/>
+        For file-format-specific options, see the related methods in <code>DataStreamReader</code>
+        (<a href="api/scala/index.html#org.apache.spark.sql.streaming.DataStreamReader">Scala</a>/<a href="api/java/org/apache/spark/sql/streaming/DataStreamReader.html">Java</a>/<a href="api/python/pyspark.sql.html#pyspark.sql.streaming.DataStreamReader">Python</a>).
+        E.g. for "parquet" format options see <code>DataStreamReader.parquet()</code></td>
+    <td>Yes</td>
+    <td>Supports regular expressions, but does not support multiple comma-separated paths/expressions.</td>
```
Comment (zsxwing) [inline docs/structured-streaming-programming-guide.md]: nit: `regular expressions` -> `glob paths`

**Comment by zsxwing** [comment]
LGTM

**Comment by david-weiluo-ren** [comment]
@tdas 
It says “However, note that all of the operations applicable on static DataFrames/Datasets are not supported in streaming DataFrames/Datasets yet” in https://spark.apache.org/docs/2.1.0/structured-streaming-programming-guide.html#unsupported-operations

I think it should be “not all of the operations …. are supported in … yet” instead of “all of the operations … are not supported in … yet"

**Comment by tdas** [comment]
@david-weiluo-ren yeah the wording can be better. maybe "all of the operations ... are not yet supported"

File: docs/structured-streaming-programming-guide.md [hunk 4]
```diff
@@ -433,6 +433,51 @@ In Spark 2.0, there are a few built-in sources.
 
   - **Socket source (for testing)** - Reads UTF8 text data from a socket connection. The listening server socket is at the driver. Note that this should be used only for testing as this does not provide end-to-end fault-tolerance guarantees. 
 
+Here are all the source details.
+
+<table class="table">
+  <tr>
+    <th>Source</th>
+    <th>Options</th>
+    <th>Fault-tolerant</th>
```
Comment (tdas) [inline docs/structured-streaming-programming-guide.md]: I dont want to make the table heading a link, but I will do something.

File: docs/structured-streaming-programming-guide.md [hunk 5]
```diff
@@ -954,49 +1014,93 @@ There are a few types of built-in output sinks.
...
-- **Memory sink (for debugging)** - The output is stored in memory as an in-memory table.  Both, Append and Complete output modes, are supported. This should be used for debugging purposes on low data volumes as the entire output is collected and stored in the driver's memory after every trigger.
+{% highlight scala %}
+writeStream
+    .format("console")
+    .start()
+{% endhighlight %}
+
+- **Memory sink (for debugging)** - The output is stored in memory as an in-memory table.
+Both, Append and Complete output modes, are supported. This should be used for debugging purposes
+on low data volumes as the entire output is collected and stored in the driver's memory after
```
Comment (tdas) [inline docs/structured-streaming-programming-guide.md]: I kind of agree it repetitive, but i dont want people to miss this.

**Comment by tdas** [comment]
Thank you very much @thomaso-mirodin @david-weiluo-ren @zsxwing 
I have addressed your comments.

**Comment by spoddutur** [comment]
Hi TD, 

As part of 2.1.0 release, Kafka as a source is added. 
SPARK-17346: Kafka 0.10 support in Structured Streaming. 
Wondering if kinesis support will be added in future. If yes, When can we expect it?

Reason for asking kinesis support is, we use kinesis spark streaming with spark 1.6 as of now and are planning to upgrade to Spark 2 Structured Streaming. So, kinda eager to know when can we expect kinesis support in StructuredStreaming.

Thanks in Advance,
Sruthi

**Pull request closed without merging**